File: packages/cli/src/commands/history.ts
import { resolvePathFromUserInput } from "@runxhq/core/config";

import { runNativeRunxJson } from "../native-runx.js";
import { relativeTime, shortId, statusIcon, theme } from "../ui.js";

export interface HistoryCommandArgs {
  readonly receiptDir?: string;
  readonly historyQuery?: string;
  readonly historySkill?: string;
  readonly historyStatus?: string;
  readonly historySource?: string;
  readonly historyActor?: string;
  readonly historyArtifactType?: string;
  readonly historySince?: string;
  readonly historyUntil?: string;
}

export interface LocalReceiptSummary {
  readonly id: string;
  readonly kind: string;
  readonly name: string;
  readonly status: string;
  readonly sourceType?: string;
  readonly disposition?: string;
  readonly outcomeState?: string;
  readonly startedAt?: string;
  readonly completedAt?: string;
  readonly actors?: readonly string[];
  readonly artifactTypes?: readonly string[];
  readonly runnerProvider?: string;
  readonly approval?: {
    readonly decision?: string;
    readonly gateType?: string;
  };
  readonly lineage?: {
    readonly kind: string;
    readonly sourceRunId: string;
  };
  readonly verification?: {
    readonly status?: string;
    readonly reason?: string;
  };
  readonly ledgerVerification?: {
    readonly status?: string;
    readonly reason?: string;
  };
  readonly harnessId?: string;
  readonly harnessState?: string;
  readonly harnessSealSummary?: string;
}

export interface PausedRunSummary {
  readonly id: string;
  readonly kind: string;
  readonly name: string;
  readonly status: string;
  readonly selectedRunner?: string;
  readonly stepIds: readonly string[];
  readonly stepLabels: readonly string[];
  readonly ledgerVerification?: {
    readonly status?: string;
    readonly reason?: string;
  };
}

export async function handleHistoryCommand(
  parsed: HistoryCommandArgs,
  env: NodeJS.ProcessEnv,
): Promise<{ readonly receipts: readonly LocalReceiptSummary[]; readonly pendingRuns: readonly PausedRunSummary[] }> {
  const args = ["history"];
  if (parsed.historyQuery) args.push(parsed.historyQuery);
  pushOptionalFlag(args, "--receipt-dir", parsed.receiptDir ? resolvePathFromUserInput(parsed.receiptDir, env) : undefined);
  pushOptionalFlag(args, "--skill", parsed.historySkill);
  pushOptionalFlag(args, "--status", parsed.historyStatus);
  pushOptionalFlag(args, "--source", parsed.historySource);
  pushOptionalFlag(args, "--actor", parsed.historyActor);
  pushOptionalFlag(args, "--artifact-type", parsed.historyArtifactType);
  pushOptionalFlag(args, "--since", parsed.historySince);
  pushOptionalFlag(args, "--until", parsed.historyUntil);
  args.push("--json");
  return normalizeHistoryProjection(await runNativeRunxJson(args, { env }));
}

function normalizeHistoryProjection(value: unknown): {
  readonly receipts: readonly LocalReceiptSummary[];
  readonly pendingRuns: readonly PausedRunSummary[];
} {
  const projection = asRecord(value);
  if (!projection) {
    throw new Error("native runx history returned a non-object payload.");
  }
  return {
    receipts: arrayValue(projection.receipts).map(normalizeHistoryReceipt),
    pendingRuns: arrayValue(projection.pendingRuns).map(normalizePausedRun),
  };
}

function normalizeHistoryReceipt(value: unknown): LocalReceiptSummary {
  const receipt = asRecord(value);
  if (!receipt || typeof receipt.id !== "string" || typeof receipt.name !== "string" || typeof receipt.status !== "string") {
    throw new Error("native runx history returned an invalid receipt entry.");
  }
  const verification = asRecord(receipt.verification);
  return {
    id: receipt.id,
    kind: stringValue(receipt.source_type) ?? "receipt",
    name: receipt.name,
    status: receipt.status,
    sourceType: stringValue(receipt.source_type),
    startedAt: stringValue(receipt.created_at),
    actors: stringArray(receipt.actors),
    artifactTypes: stringArray(receipt.artifact_types),
    verification: verification ? { status: stringValue(verification.status) } : undefined,
    harnessId: stringValue(receipt.harness_id),
    harnessState: stringValue(receipt.harness_state),
    harnessSealSummary: stringValue(receipt.summary),
  };
}

function normalizePausedRun(value: unknown): PausedRunSummary {
  const run = asRecord(value);
  if (!run || typeof run.id !== "string" || typeof run.name !== "string" || typeof run.kind !== "string" || typeof run.status !== "string") {
    throw new Error("native runx history returned an invalid pending run entry.");
  }
  const ledgerVerification = asRecord(run.ledgerVerification);
  return {
    id: run.id,
    name: run.name,
    kind: run.kind,
    status: run.status === "paused" ? "needs_agent" : run.status,
    selectedRunner: stringValue(run.selectedRunner),
    stepIds: stringArray(run.stepIds),
    stepLabels: stringArray(run.stepLabels),
    ledgerVerification: ledgerVerification
      ? {
          status: stringValue(ledgerVerification.status),
          reason: stringValue(ledgerVerification.reason),
        }
      : undefined,
  };
}

function pushOptionalFlag(args: string[], flag: string, value: string | undefined): void {
  if (value !== undefined && value.length > 0) {
    args.push(flag, value);
  }
}

function asRecord(value: unknown): Readonly<Record<string, unknown>> | undefined {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? value as Readonly<Record<string, unknown>>
    : undefined;
}

function arrayValue(value: unknown): readonly unknown[] {
  return Array.isArray(value) ? value : [];
}

function stringValue(value: unknown): string | undefined {
  return typeof value === "string" ? value : undefined;
}

function stringArray(value: unknown): readonly string[] {
  return Array.isArray(value) ? value.filter((entry): entry is string => typeof entry === "string") : [];
}

export function renderHistory(
  receipts: readonly LocalReceiptSummary[],
  env: NodeJS.ProcessEnv = process.env,
  query?: string,
  pendingRuns: readonly PausedRunSummary[] = [],
): string {
  const t = theme(undefined, env);
  const totalCount = receipts.length + pendingRuns.length;
  if (totalCount === 0) {
    return query
      ? `\n  ${t.dim}No receipts matched ${t.cyan}${query}${t.reset}${t.dim}.${t.reset}\n  ${t.dim}Try ${t.cyan}runx history${t.reset}${t.dim} to see every local run.${t.reset}\n\n`
      : `\n  ${t.dim}No receipts yet. Try a run first:${t.reset}\n  ${t.cyan}runx skill <skill-dir> --json${t.reset}\n  ${t.cyan}runx list skills${t.reset}\n\n`;
  }
  const now = Date.now();
  const allNames = [...receipts.map((r) => r.name), ...pendingRuns.map((r) => r.name)];
  const nameWidth = Math.min(32, Math.max(...allNames.map((name) => name.length)));
  const lines: string[] = [""];
  const summary = pendingRuns.length > 0
    ? `${receipts.length} receipt(s), ${pendingRuns.length} needs_agent`
    : `${totalCount} receipt(s)`;
  lines.push(`  ${t.bold}history${t.reset}${query ? `  ${t.dim}· ${query}${t.reset}` : ""}  ${t.dim}${summary}${t.reset}`);
  lines.push("");
  for (const pending of pendingRuns) {
    const name = pending.name.padEnd(nameWidth);
    const id = shortId(pending.id);
    const stepLabel = pending.stepLabels[0] ?? pending.stepIds[0] ?? "—";
    lines.push(
      `  ${t.cyan}◇${t.reset}  ${t.bold}${name}${t.reset}  ${t.dim}${pending.status.padEnd(16)}${t.reset}  ${t.dim}${stepLabel.padEnd(10)}${t.reset}  ${t.dim}${"".padEnd(10)}${t.reset}  ${t.dim}${id}${t.reset}`,
    );
  }
  for (const receipt of receipts) {
    const icon = statusIcon(receipt.status, t);
    const name = receipt.name.padEnd(nameWidth);
    const when = receipt.startedAt ? relativeTime(receipt.startedAt, now) : "";
    const source = receipt.sourceType ?? receipt.kind;
    const id = shortId(receipt.id);
    const verification = formatHistoryVerification(receipt);
    lines.push(
      `  ${icon}  ${t.bold}${name}${t.reset}  ${t.dim}${source.padEnd(16)}${t.reset}  ${t.dim}${verification.padEnd(16)}${t.reset}  ${t.dim}${when.padEnd(10)}${t.reset}  ${t.dim}${id}${t.reset}`,
    );
    const harnessStatus = formatHarnessHistoryStatus(receipt);
    if (harnessStatus) {
      lines.push(`     ${t.dim}${harnessStatus}${t.reset}`);
    }
  }
  lines.push("");
  if (pendingRuns.length > 0) {
    lines.push(`  ${t.dim}next${t.reset}  runx skill <same-skill-ref> --run-id <run-id> --answers answers.json  ${t.dim}or${t.reset}  runx history --json`);
  } else {
    lines.push(`  ${t.dim}next${t.reset}  runx history --json`);
  }
  lines.push("");
  return lines.join("\n");
}

function formatHarnessHistoryStatus(receipt: LocalReceiptSummary): string | undefined {
  if (!receipt.harnessState && !receipt.harnessSealSummary && !receipt.harnessId) {
    return undefined;
  }
  const parts = [
    receipt.harnessId ? `harness ${receipt.harnessId}` : "harness",
    receipt.harnessState,
    receipt.harnessSealSummary,
  ].filter((value): value is string => Boolean(value));
  return parts.join(" · ");
}

function formatHistoryVerification(receipt: LocalReceiptSummary): string {
  const signature = receipt.verification?.status ?? "unknown";
  const ledger = receipt.ledgerVerification?.status ?? "unknown";
  return `${signature}/${ledger}`;
}

File: crates/runx-cli/src/history.rs
use std::collections::BTreeMap;
use std::env;
use std::ffi::OsString;
use std::fmt;
use std::path::{Path, PathBuf};

use runx_runtime::journal::{
    HistoryFilter, JournalProjectionError, list_local_history, list_local_history_with_policy,
};
use runx_runtime::{
    Ed25519ReceiptVerifier, LocalReceiptStore, ReceiptPathInputs, RuntimeReceiptConfig,
    RuntimeReceiptSignaturePolicy, resolve_receipt_path,
};

// rust-style-allow: large-file because the native history CLI slice keeps
// parsing, rendering, and CLI parity tests together until the rest of the Rust
// command routing settles.
#[derive(Debug)]
pub enum HistoryCliError {
    InvalidArgs(String),
    InvalidReceiptVerifier(String),
    Projection(JournalProjectionError),
    Serialize(serde_json::Error),
}

impl fmt::Display for HistoryCliError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidArgs(message) => formatter.write_str(message),
            Self::InvalidReceiptVerifier(message) => formatter.write_str(message),
            Self::Projection(error) => write!(formatter, "{error}"),
            Self::Serialize(error) => write!(formatter, "failed to serialize history: {error}"),
        }
    }
}

impl std::error::Error for HistoryCliError {}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct HistoryCliResult {
    pub output: String,
    pub error_is_usage: bool,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
struct ParsedHistoryArgs {
    receipt_dir: Option<PathBuf>,
    query: Option<String>,
    filter: HistoryFilter,
    json: bool,
}

pub fn run_history_command(
    args: &[OsString],
    env: &BTreeMap<String, String>,
    cwd: &Path,
) -> Result<HistoryCliResult, HistoryCliError> {
    let parsed = parse_history_args(args)?;
    let receipt_config = RuntimeReceiptConfig::default();
    let resolved = resolve_receipt_path(ReceiptPathInputs {
        explicit_dir: parsed.receipt_dir.as_deref(),
        runtime_config: Some(&receipt_config),
        env,
        cwd,
    });
    let store = LocalReceiptStore::new(&resolved.path);
    let verifier = history_production_verifier(env)?;
    let history = if let Some(verifier) = verifier.as_ref() {
        list_local_history_with_policy(
            &store,
            &resolved.workspace_base,
            &resolved.project_runx_dir,
            &parsed.filter,
            RuntimeReceiptSignaturePolicy::production(verifier),
        )
    } else {
        list_local_history(
            &store,
            &resolved.workspace_base,
            &resolved.project_runx_dir,
            &parsed.filter,
        )
    }
    .map_err(HistoryCliError::Projection)?;
    let output = if parsed.json {
        format!(
            "{}\n",
            serde_json::to_string_pretty(&history).map_err(HistoryCliError::Serialize)?
        )
    } else {
        render_history(&history, parsed.query.as_deref())
    };
    Ok(HistoryCliResult {
        output,
        error_is_usage: false,
    })
}

const RUNX_RECEIPT_VERIFY_KID_ENV: &str = "RUNX_RECEIPT_VERIFY_KID";
const RUNX_RECEIPT_VERIFY_ED25519_PUBLIC_KEY_BASE64_ENV: &str =
    "RUNX_RECEIPT_VERIFY_ED25519_PUBLIC_KEY_BASE64";

fn history_production_verifier(
    env: &BTreeMap<String, String>,
) -> Result<Option<Ed25519ReceiptVerifier>, HistoryCliError> {
    let kid = non_empty_env(env, RUNX_RECEIPT_VERIFY_KID_ENV);
    let public_key = non_empty_env(env, RUNX_RECEIPT_VERIFY_ED25519_PUBLIC_KEY_BASE64_ENV);
    match (kid, public_key) {
        (None, None) => Ok(None),
        (Some(kid), Some(public_key)) => Ed25519ReceiptVerifier::from_public_key_base64(
            kid.to_owned(),
            public_key,
        )
        .map(Some)
        .map_err(|_| {
            HistoryCliError::InvalidReceiptVerifier(format!(
                "{RUNX_RECEIPT_VERIFY_ED25519_PUBLIC_KEY_BASE64_ENV} is not valid Ed25519 public key material"
            ))
        }),
        _ => Err(HistoryCliError::InvalidReceiptVerifier(format!(
            "{RUNX_RECEIPT_VERIFY_KID_ENV} and {RUNX_RECEIPT_VERIFY_ED25519_PUBLIC_KEY_BASE64_ENV} must be set together"
        ))),
    }
}

fn non_empty_env<'a>(env: &'a BTreeMap<String, String>, key: &str) -> Option<&'a str> {
    env.get(key)
        .map(String::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
}

pub fn env_map() -> BTreeMap<String, String> {
    env::vars().collect()
}

// rust-style-allow: long-function because this mirrors the public history CLI
// flag grammar in one parser during the hard cutover.
fn parse_history_args(args: &[OsString]) -> Result<ParsedHistoryArgs, HistoryCliError> {
    if args.first().and_then(|arg| arg.to_str()) != Some("history") {
        return Err(HistoryCliError::InvalidArgs(
            "internal error: history dispatcher received non-history command".to_owned(),
        ));
    }
    let mut parsed = ParsedHistoryArgs::default();
    let mut positionals = Vec::new();
    let mut index = 1;
    while index < args.len() {
        let token = os_arg(args, index)?;
        if !token.starts_with("--") {
            positionals.push(token.to_owned());
            index += 1;
            continue;
        }
        let (flag, inline_value) = split_flag(token);
        match flag {
            "--json" => {
                if inline_value.is_some() {
                    return Err(invalid_args("--json does not take a value"));
                }
                parsed.json = true;
                index += 1;
            }
            "--receipt-dir" => {
                let (value, next_index) = flag_value(args, index, flag, inline_value)?;
                parsed.receipt_dir = Some(PathBuf::from(value));
                index = next_index;
            }
            "--skill" => {
                let (value, next_index) = flag_value(args, index, flag, inline_value)?;
                parsed.filter.skill = Some(value);
                index = next_index;
            }
            "--status" => {
                let (value, next_index) = flag_value(args, index, flag, inline_value)?;
                parsed.filter.status = Some(value);
                index = next_index;
            }
            "--source" => {
                let (value, next_index) = flag_value(args, index, flag, inline_value)?;
                parsed.filter.source = Some(value);
                index = next_index;
            }
            "--actor" => {
                let (value, next_index) = flag_value(args, index, flag, inline_value)?;
                parsed.filter.actor = Some(value);
                index = next_index;
            }
            "--artifact-type" | "--artifact_type" | "--artifactType" => {
                let (value, next_index) = flag_value(args, index, flag, inline_value)?;
                parsed.filter.artifact_type = Some(value);
                index = next_index;
            }
            "--since" => {
                let (value, next_index) = flag_value(args, index, flag, inline_value)?;
                parsed.filter.since = Some(value);
                index = next_index;
            }
            "--until" => {
                let (value, next_index) = flag_value(args, index, flag, inline_value)?;
                parsed.filter.until = Some(value);
                index = next_index;
            }
            "--limit" => {
                let (value, next_index) = flag_value(args, index, flag, inline_value)?;
                parsed.filter.limit = Some(value.parse().map_err(|_| {
                    HistoryCliError::InvalidArgs(format!("invalid --limit value '{value}'"))
                })?);
                index = next_index;
            }
            _ => {
                return Err(HistoryCliError::InvalidArgs(format!(
                    "unknown history flag {flag}"
                )));
            }
        }
    }
    parsed.query = (!positionals.is_empty()).then(|| positionals.join(" "));
    parsed.filter.query = parsed.query.clone();
    Ok(parsed)
}

fn render_history(
    history: &runx_runtime::journal::LocalHistoryProjection,
    query: Option<&str>,
) -> String {
    let total = history.receipts.len() + history.pending_runs.len();
    if total == 0 {
        if let Some(query) = query {
            return format!(
                "\n  No receipts matched {query}.\n  Try runx history to see every local run.\n\n"
            );
        }
        return "\n  No receipts yet. Try a run first:\n  runx skill <skill-dir> --json\n  runx harness <fixture.yaml> --json\n\n"
            .to_owned();
    }
    let mut lines = Vec::new();
    lines.push(String::new());
    if history.pending_runs.is_empty() {
        lines.push(format!("  history  {} receipt(s)", history.receipts.len()));
    } else {
        lines.push(format!(
            "  history  {} receipt(s), {} needs_agent",
            history.receipts.len(),
            history.pending_runs.len()
        ));
    }
    lines.push(String::new());
    for pending in &history.pending_runs {
        let step = pending
            .step_labels
            .first()
            .or_else(|| pending.step_ids.first())
            .map_or("", String::as_str);
        lines.push(format!(
            "  *  {}  needs_agent  {}  {}",
            pending.name,
            step,
            short_id(&pending.id)
        ));
    }
    for receipt in &history.receipts {
        lines.push(format!(
            "  {}  {}  {}  {}",
            receipt.status,
            receipt.name,
            receipt.verification.status,
            short_id(&receipt.id)
        ));
    }
    lines.push(String::new());
    if history.pending_runs.is_empty() {
        lines.push("  next  runx history <receipt-id> --json".to_owned());
    } else {
        lines.push(
            "  next  rerun the same runx skill <path> with --run-id and --answers".to_owned(),
        );
    }
    lines.push(String::new());
    lines.join("\n")
}

fn short_id(value: &str) -> &str {
    value.get(..12).unwrap_or(value)
}

fn os_arg(args: &[OsString], index: usize) -> Result<&str, HistoryCliError> {
    args.get(index)
        .and_then(|arg| arg.to_str())
        .ok_or_else(|| HistoryCliError::InvalidArgs("history arguments must be UTF-8".to_owned()))
}

fn split_flag(token: &str) -> (&str, Option<&str>) {
    token
        .split_once('=')
        .map_or((token, None), |(flag, value)| (flag, Some(value)))
}

fn flag_value(
    args: &[OsString],
    index: usize,
    flag: &str,
    inline_value: Option<&str>,
) -> Result<(String, usize), HistoryCliError> {
    if let Some(value) = inline_value {
        return Ok((value.to_owned(), index + 1));
    }
    let value = os_arg(args, index + 1)
        .map_err(|_| HistoryCliError::InvalidArgs(format!("{flag} requires a value")))?;
    if value.starts_with("--") {
        return Err(HistoryCliError::InvalidArgs(format!(
            "{flag} requires a value"
        )));
    }
    Ok((value.to_owned(), index + 2))
}

fn invalid_args(message: &str) -> HistoryCliError {
    HistoryCliError::InvalidArgs(message.to_owned())
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::io;

    use super::*;
    use runx_contracts::ReceiptIssuerType;
    use runx_runtime::receipts::step_receipt_with_signature_policy;
    use runx_runtime::{Ed25519ReceiptSigner, InvocationStatus, RuntimeError, SkillOutput};

    #[test]
    fn parses_history_args_without_comparing_against_runtime_constants() -> Result<(), io::Error> {
        let parsed = parse_history_args(&[
            "history".into(),
            "sourcey".into(),
            "--skill".into(),
            "source".into(),
            "--status=needs_agent".into(),
            "--artifact-type".into(),
            "artifact".into(),
            "--json".into(),
        ])
        .map_err(|error| io::Error::other(error.to_string()))?;

        assert_eq!(parsed.query.as_deref(), Some("sourcey"));
        assert_eq!(parsed.filter.skill.as_deref(), Some("source"));
        assert_eq!(parsed.filter.status.as_deref(), Some("needs_agent"));
        assert_eq!(parsed.filter.artifact_type.as_deref(), Some("artifact"));
        assert!(parsed.json);
        Ok(())
    }

    #[test]
    // rust-style-allow: long-function because the CLI execute oracle test keeps
    // its ledger fixture, command invocation, and typed output assertions in
    // one place so the parity case remains readable.
    fn executes_history_json_against_cli_parity_oracle() -> Result<(), io::Error> {
        let temp = tempfile_dir()?;
        let receipt_dir = temp.join("receipts");
        fs::create_dir_all(receipt_dir.join("ledgers"))?;
        fs::write(
            receipt_dir
                .join("ledgers")
                .join("gx_needs_agent_oracle.jsonl"),
            format!(
                "{}\n{}\n",
                r#"{"entry":{"type":"run_event","version":"1","data":{"kind":"run_started","status":"started","step_id":null,"detail":{}},"meta":{"artifact_id":"ax_start","run_id":"gx_needs_agent_oracle","step_id":null,"producer":{"skill":"sourcey","runner":"graph"},"created_at":"2026-04-28T01:00:00.000Z","hash":"sha256:start","size_bytes":2,"parent_artifact_id":null,"receipt_id":null,"redacted":false}}}"#,
                r#"{"entry":{"type":"run_event","version":"1","data":{"kind":"step_waiting_resolution","status":"waiting","step_id":"discover","detail":{"request_ids":["agent_step.test-step.output"],"resolution_kinds":["agent_act"],"step_ids":["discover"],"step_labels":["inspect repo"],"inputs":{},"selected_runner":"agent-step"}},"meta":{"artifact_id":"ax_wait","run_id":"gx_needs_agent_oracle","step_id":"discover","producer":{"skill":"sourcey","runner":"graph"},"created_at":"2026-04-28T01:00:00.000Z","hash":"sha256:wait","size_bytes":2,"parent_artifact_id":null,"receipt_id":null,"redacted":false}}}"#
            ),
        )?;
        let oracle: CliParityOracle = serde_json::from_str(include_str!(
            "../../../fixtures/cli-parity/cases/oracle.json"
        ))
        .map_err(|error| io::Error::other(error.to_string()))?;
        let execute_case = oracle
            .cases
            .iter()
            .find(|case| case.id == "history.execute")
            .ok_or_else(|| {
                io::Error::new(
                    io::ErrorKind::InvalidData,
                    "missing history.execute oracle case",
                )
            })?;

        let mut env = BTreeMap::new();
        env.insert("RUNX_CWD".to_owned(), temp.to_string_lossy().to_string());
        let result = run_history_command(
            &[
                "history".into(),
                "--receipt-dir".into(),
                receipt_dir.into_os_string(),
                "--json".into(),
            ],
            &env,
            &temp,
        )
        .map_err(|error| io::Error::other(error.to_string()))?;
        let output: HistoryOutput = serde_json::from_str(&result.output)
            .map_err(|error| io::Error::other(error.to_string()))?;
        let first_pending_run = output.pending_runs.first().ok_or_else(|| {
            io::Error::new(
                io::ErrorKind::InvalidData,
                "history output has no pending run",
            )
        })?;

        assert_eq!(
            output.pending_runs.len(),
            execute_case.expect.pending_runs as usize
        );
        assert_eq!(
            first_pending_run.id,
            execute_case.expect.first_pending_run_id
        );
        assert_eq!(
            first_pending_run.status,
            execute_case.expect.first_pending_run_status
        );
        assert_eq!(
            first_pending_run.selected_runner,
            Some("agent-step".to_owned())
        );
        Ok(())
    }

    #[test]
    fn history_json_reports_production_verified_receipts_when_verifier_env_is_configured()
    -> Result<(), io::Error> {
        let temp = tempfile_dir()?;
        let receipt_dir = temp.join("receipts");
        let signer = fixture_signer().map_err(|error| io::Error::other(error.to_string()))?;
        let receipt = production_signed_receipt(&signer)
            .map_err(|error| io::Error::other(error.to_string()))?;
        let store = LocalReceiptStore::new(&receipt_dir);
        let verifier = Ed25519ReceiptVerifier::new([signer.production_key()]);
        store
            .write_receipt_with_policy(
                &receipt,
                RuntimeReceiptSignaturePolicy::production(&verifier),
            )
            .map_err(|error| io::Error::other(error.to_string()))?;

        let mut env = BTreeMap::new();
        env.insert("RUNX_CWD".to_owned(), temp.to_string_lossy().to_string());
        env.insert(
            RUNX_RECEIPT_VERIFY_KID_ENV.to_owned(),
            FIXTURE_KID.to_owned(),
        );
        env.insert(
            RUNX_RECEIPT_VERIFY_ED25519_PUBLIC_KEY_BASE64_ENV.to_owned(),
            base64_standard(signer.public_key()),
        );

        let result = run_history_command(
            &[
                "history".into(),
                receipt.id.to_string().into(),
                "--receipt-dir".into(),
                receipt_dir.into_os_string(),
                "--json".into(),
            ],
            &env,
            &temp,
        )
        .map_err(|error| io::Error::other(error.to_string()))?;
        let output: HistoryOutput = serde_json::from_str(&result.output)
            .map_err(|error| io::Error::other(error.to_string()))?;
        let first_receipt = output.receipts.first().ok_or_else(|| {
            io::Error::new(io::ErrorKind::InvalidData, "history output has no receipt")
        })?;

        assert_eq!(first_receipt.id, receipt.id.to_string());
        assert_eq!(first_receipt.verification.status, "verified");
        Ok(())
    }

    #[derive(serde::Deserialize)]
    struct CliParityOracle {
        cases: Vec<CliParityCase>,
    }

    #[derive(serde::Deserialize)]
    struct CliParityCase {
        id: String,
        #[serde(default)]
        expect: CliParityExpectation,
    }

    #[derive(Default, serde::Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct CliParityExpectation {
        #[serde(default)]
        pending_runs: u64,
        #[serde(default)]
        first_pending_run_id: String,
        #[serde(default)]
        first_pending_run_status: String,
    }

    #[derive(serde::Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct HistoryOutput {
        receipts: Vec<HistoryReceipt>,
        pending_runs: Vec<HistoryPendingRun>,
    }

    #[derive(serde::Deserialize)]
    struct HistoryReceipt {
        id: String,
        verification: HistoryReceiptVerification,
    }

    #[derive(serde::Deserialize)]
    struct HistoryReceiptVerification {
        status: String,
    }

    #[derive(serde::Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct HistoryPendingRun {
        id: String,
        status: String,
        selected_runner: Option<String>,
    }

    fn tempfile_dir() -> Result<PathBuf, io::Error> {
        let path = std::env::temp_dir().join(format!(
            "runx-cli-history-{}-{}",
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map_err(|error| io::Error::other(error.to_string()))?
                .as_nanos()
        ));
        fs::create_dir_all(&path)?;
        Ok(path)
    }

    const FIXTURE_KID: &str = "runx-cli-prod-history-fixture-key";
    const FIXTURE_SEED: [u8; 32] = [0x42; 32];

    fn fixture_signer() -> Result<Ed25519ReceiptSigner, runx_runtime::RuntimeReceiptSigningError> {
        Ed25519ReceiptSigner::from_seed(FIXTURE_KID, ReceiptIssuerType::Local, &FIXTURE_SEED)
    }

    fn production_signed_receipt(
        signer: &Ed25519ReceiptSigner,
    ) -> Result<runx_contracts::Receipt, RuntimeError> {
        let verifier = Ed25519ReceiptVerifier::new([signer.production_key()]);
        let output = SkillOutput {
            status: InvocationStatus::Success,
            stdout:
                r#"{"artifact":{"artifact_id":"artifact_cli_history","artifact_type":"artifact"}}"#
                    .to_owned(),
            stderr: String::new(),
            exit_code: Some(0),
            duration_ms: 10,
            metadata: BTreeMap::new(),
        };
        step_receipt_with_signature_policy(
            "cli-history",
            "production-verified",
            1,
            &output,
            "2026-05-25T00:00:00Z",
            RuntimeReceiptSignaturePolicy::production_signing(signer, &verifier),
        )
    }

    fn base64_standard(bytes: &[u8]) -> String {
        const TABLE: &[u8; 64] =
            b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
        let mut encoded = String::with_capacity(bytes.len().div_ceil(3) * 4);
        for chunk in bytes.chunks(3) {
            let first = chunk[0];
            let second = chunk.get(1).copied().unwrap_or(0);
            let third = chunk.get(2).copied().unwrap_or(0);
            let combined = ((first as u32) << 16) | ((second as u32) << 8) | third as u32;
            encoded.push(TABLE[((combined >> 18) & 0x3f) as usize] as char);
            encoded.push(TABLE[((combined >> 12) & 0x3f) as usize] as char);
            if chunk.len() > 1 {
                encoded.push(TABLE[((combined >> 6) & 0x3f) as usize] as char);
            } else {
                encoded.push('=');
            }
            if chunk.len() > 2 {
                encoded.push(TABLE[(combined & 0x3f) as usize] as char);
            } else {
                encoded.push('=');
            }
        }
        encoded
    }
}
